### PR TITLE
fix: RelativeLayoutCache must hold disk dtype

### PR DESCRIPTION
**Pull request by danking**

### CODE REVIEW

File: vortex-dtype/src/dtype.rs [hunk 1]
```diff
@@ -169,6 +169,19 @@ impl StructDType {
...
+        self.maybe_field(field)
+            .ok_or_else(|| vortex_err!("expected field ({}) to exist", field))
+    }
+
+    pub fn maybe_field(&self, field: &Field) -> Option<(usize, Arc<str>, &DType)> {
+        let index = match field {
+            Field::Name(name) => self.find_name(name)?,
+            Field::Index(index) => *index,
+        };
+        Some((index, self.names[index].clone(), &self.dtypes[index]))
```
Comment (gatesn) [inline vortex-dtype/src/dtype.rs]: This panics if Field::Index(index) is OOB?

Comment (danking) [reply vortex-dtype/src/dtype.rs]: Done and unified the logic into this method from below.

**Review by lwwmanning — APPROVED** [review]

File: vortex-serde/src/layouts/read/layouts.rs [hunk 2]
```diff
@@ -189,26 +189,31 @@ impl LayoutReader for ColumnLayout {
...
+                            })
+                            .collect::<VortexResult<Vec<_>>>()?;
+                        (struct_dtype.names().clone(), layouts)
+                    }
+                    Projection::Flat(projected_fields) => {
+                        let mut names = Vec::with_capacity(projected_fields.len());
+                        let mut layouts = Vec::with_capacity(projected_fields.len());
+                        for projected_field in projected_fields {
+                            let FieldInfo { index, name, dtype } =
+                                struct_dtype.field_info(projected_field)?;
```
Comment (robert3005) [inline vortex-serde/src/layouts/read/layouts.rs:207]: yes this, I had a change that just projected dtype and the re resolved fields here

**Review by robert3005 — APPROVED** [review]

File: vortex-serde/src/layouts/read/builder.rs [hunk 3]
```diff
@@ -107,7 +107,7 @@ impl<R: VortexReadAt> LayoutReaderBuilder<R> {
 
         let data_reader = footer.layout(
             scan.clone(),
-            RelativeLayoutCache::new(message_cache.clone(), projected_dtype.clone()),
+            RelativeLayoutCache::new(message_cache.clone(), footer.dtype()?),
```
Comment (robert3005) [inline vortex-serde/src/layouts/read/builder.rs]: you need same fix on line 123

**Pull request merged**